### ISSUE TITLE
Blazor Server - NavigationManager push to history

### ISSUE THREAD

**Issue by lofcz** [dotnet/aspnetcore#51042]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

There is no clear way to implement a user-friendly list & detail page pattern. Consider an application consisting of two pages: 
```
/articles
/article/{id}
```

The `/articles` endpoint enumerates articles, and then the user opens one. After the user is finished with the article, they navigate back to the `/articles` endpoint. This will result in the `/articles` page fetching and rendering the articles again and the vertical scroll position of the page incorrectly reset. 

Now the user has to scroll to the place where they left off again. This is a bad user experience.

Currently, our options are:
1. We can remedy the scrolling issue by storing the scroll position when the navigation is about to happen and restoring it after we are brought back to `/articles` via `NavigationManager.RegisterLocationChangingHandler`, `NavigationManager.LocationChanged`, and some JS.

**However, this approach is flawed. In order to scroll back to the correct position we need to wait for the articles to be fetched, rendered and only after that we can scroll.** 

2. We can exploit the fact that navigation where only the query string is changed eg. from `/articles` to `/articles?detail={id}` doesn't result in the original content being lost and with `OnParametersSet` implement a switch where we either render the articles or the requested detail. We still need to fix the scroll the same way as in (1) but this time we avoid fetching the articles again hence removing the delay and improving the end-user experience.

**However, this way we make a compromise with the URLs. This isn't a sustainable way to solve the issue.**

3. We avoid Blazor's navigation manager altogether and solve the issue with a combination of JS's `history` API calls. This is fairly hard to get working correctly if we want to intercept only certain navigation events (considering there could be more endpoints and we don't want to get these involved). We can hook `popstate`, `href`s being clicked, and so on before Blazor's JS is loaded so we can snatch these events as needed.

**Again, this is pretty hard and the framework is a hindrance actively fighting against you at that point.**

Now consider the articles could have a parameter attached to them, say a number of likes and we want that number to be updated when we navigate back from an article detail to the list. 

A real-world example to grasp this concept is simple: navigate to Reddit, scroll a little, open a post, and navigate back. Observe the likes updated, and the scroll position kept intact.

To do this, the ideal way is to hide the content of a container listing the articles when we navigate to the detail of an article, rendering the article itself in another container such as:

```
<div class="articles"></div>
<div class="articleDetail"></div>
```

This can be done in quite a few ways in Blazor, however, we can't easily change the URL while preserving the original content, essentially what `history.pushState` does. If we call this function ourselves, Blazor will avoid intercepting the navigation, see https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/Services/NavigationManager.ts#L184 as to why and how.

### Describe the solution you'd like

The clearest way to solve this would be to introduce a new API:

```diff
+NavigationManager.PushState(string url, string? historyEntryState = null)
```

**Comment by lofcz** [comment]
@SteveSandersonMS pinging as I can offer to implement this, however I'd like to gather feedback on the proposal first.

**Comment by SteveSandersonMS** [comment]
Thanks for offering. It sounds like a potentially complex feature so whether or not we would accept it depends a lot on how simple the implementation could be, how simple any options would be that developers would need to control, and whether we can be confident enough it won't cause other problems or complexities.

If you're interested in exploring this please go ahead! Just wanted to be clear up front that if it turns out to be complex or a potential risk area then we would be less likely to take it as a core framework feature, and might want to suggest ways to plug it in as an external package instead.

**Comment by lofcz** [comment]
@SteveSandersonMS I cobbled a quick prototype together - as of now, this is implemented only for Blazor Server, Webassembly & WebView implementation is pending.

https://github.com/lofcz/aspnetcore/tree/feat-push-state

It seems working from the preliminary tests I did however I have a hard time developing E2E tests as the way `Basic test app` routing is set up doesn't support resolving the correct page after refresh if the page is in the form of `/subdir/...`

![image](https://github.com/dotnet/aspnetcore/assets/10260230/1b96b8a1-3f0b-41a0-a6bb-4ebf44152dad)
(after refreshing the select is reset)

I've considered extending the `Router` demo with my two pages representing a list of some kind and a detail. However, to test it fully I'd need to do something like:

```
1. /subdir/pushStateOverview
2. /subdir/pushStateDetail/{id}
3. user refreshes the tab
4. /subdir/pushStateDetail/{id} still has the correct content
```

Could you advise on how to proceed? Thanks.

**Comment by SteveSandersonMS** [comment]
Checking the implementation, I could use a bit of clarification of your goals. I thought your goal was to deal with scroll position restoration, but it seems like the code only deals with calling `history.pushState`.

Isn't this the same as what the framework already does if you call `NavigationManager.NavigateTo`?

**Comment by lofcz** [comment]
My intention is to be able to push to history in a way that doesn't trigger navigation and at the same time will be picked by the built-in navigation manager. Things such as restoring scroll position or other end-user-specific requests are not in the scope of the proposal I've merely illustrated how the proposed API would be typically used.
The goal here is to provide the bare minimum for users to enable implementation of such features without having to resort to listening to `popstate` event before Blazor and conditionally canceling the bubbling before Blazor reacts to the event itself.

Hence I think this proposal isn't that complicated and could be worked into a mergable PR.

**Comment by gilm0079** [comment]
I would vote for this as well. I was under the assumption that the NavigateTo would not execute navigation if I was just updating the URL with new query params, but I'm finding out that my app is going through a dispose/initialize cycle when I use NavigateTo for query param updates. I'm changing my app to use JS history.pushState in the meantime until NavigationManager natively supports silent url updates.

**Comment by SteveSandersonMS** [comment]
@gilm0079 It's hard for us to be sure whether that's the same requirement as described in this issue or not. There are lots of things you could mean based on what hosting model you're using, when this redirection would occur, and so on. To avoid confusion, would you consider posting a new issue, rather than continuing this thread, and including a minimal code example showing what you are trying to do that causes the unwanted dispose/initialize cycle? Thanks!

**Comment by gilm0079** [comment]
@SteveSandersonMS I created a new issue, linked here. There is a minimal solution I have started to work on. It is not replicating with the minimal blazor webassembly so I'll keep adding complexities until it does. Any suggestions would be helpful. Thanks!

**Comment by SteveSandersonMS** [comment]
@gilm0079 If you're saying it works OK in the default template project, but doesn't work in your more complex project, the main thing to debug is finding the exact minimal change to the default project that repros the issue. That will indicate whether it's a bug in your application code or in the framework.
